crates/sp-domains-fraud-proof/src/verification.rs

```diff
@@ -317,3 +321,142 @@ where
+        DomainHash,
+        Balance,
+    >,
+    invalid_bundle_fraud_proof: &InvalidBundlesFraudProof<ReceiptHash>,
+) -> Result<InboxedBundle<DomainHash>, VerificationError>
+where
+    CBlock: BlockT,
+    DomainHash: Into<H256> + PartialEq + Clone,
+{
+    let targeted_invalid_bundle_entry = bad_receipt
+        .inboxed_bundles
+        .get(invalid_bundle_fraud_proof.bundle_index as usize)
+        .ok_or(VerificationError::BundleNotFound)?;
+
+    let invalid_type_of_proof = match invalid_bundle_fraud_proof.proof_data {
+        ProofDataPerInvalidBundleType::OutOfRangeTx => {
+            InvalidBundleType::OutOfRangeTx(invalid_bundle_fraud_proof.mismatched_extrinsic_index)
+        }
+    };
+
+    let is_expected = if !invalid_bundle_fraud_proof.is_true_invalid_fraud_proof {
+        // `FalseInvalid`
+        // The proof trying to prove `bad_receipt_bundle`'s `invalid_bundle_type` is wrong,
+        // so the proof should contains the same `invalid_bundle_type`
+        targeted_invalid_bundle_entry.bundle == BundleValidity::Invalid(invalid_type_of_proof)
+    } else {
+        // `TrueInvalid`
+        match &targeted_invalid_bundle_entry.bundle {
+            // The proof trying to prove the bundle is invalid due to `invalid_type_of_proof` while `bad_receipt_bundle`
+            // think it is valid
+            BundleValidity::Valid(_) => true,
+            BundleValidity::Invalid(invalid_type) => {
+                // The proof trying to prove there is an invalid extrinsic that the `bad_receipt_bundle` think is valid,
+                // so the proof should point to an extrinsic that in front of the `bad_receipt_bundle`'s
+                invalid_type_of_proof.extrinsic_index() < invalid_type.extrinsic_index() ||
+                    // The proof trying to prove the invalid extrinsic can not pass a check that the `bad_receipt_bundle` think it can,
+                    // so the proof should point to the same extrinsic and a check that perform before the `bad_receipt_bundle`'s
+                    (invalid_type_of_proof.extrinsic_index() == invalid_type.extrinsic_index()
+                        && invalid_type_of_proof.checking_order() < invalid_type.checking_order())
+            }
+        }
+    };
+
+    if !is_expected {
+        return match &targeted_invalid_bundle_entry.bundle {
+            BundleValidity::Invalid(invalid_bundle_type) => {
+                Err(VerificationError::UnexpectedInvalidBundleEntry {
+                    bundle_index: invalid_bundle_fraud_proof.bundle_index,
+                    invalid_bundle_type: invalid_bundle_type.clone(),
+                })
+            }
+            BundleValidity::Valid(valid_bundle_hash) => {
+                Err(VerificationError::UnexpectedValidBundleEntry {
+                    bundle_index: invalid_bundle_fraud_proof.bundle_index,
+                    bundle_hash: valid_bundle_hash.clone().into(),
+                })
+            }
+        };
+    }
```

Comment:
> even with such many info fits into the error, we still can't tell what exactly caused the check to fail by just looking at the error

Yeah, my point is that it is not very helpful, given the error I can't locate the exact step that the check failed at. Perhaps an error that just includes `targeted_invalid_bundle_entry.bundle` and `invalid_type_of_proof`?

Reply:
@NingLin-P That was the original plan. Problem there is `BundleValidity` has generic parameter so we would need to write function just to convert `BundleValidity<Hash>` to `BundleValidity<H256>` to use it in the error or make VerificationError generic which is more trouble than its worth.  

This is the best error types I can come up with that do not require generics and able to carry information without much processing. If you want, please suggest better naming there.

Reply:
My concern is not the naming but the error not really helping in debugging, so would prefer either "write function just to convert BundleValidity<Hash> to BundleValidity<H256>" or since it can't help in debugging anyway something simpler like  `TargettingUnexpectedBundle` is better.

Reply:
Okay. I would prefer including as much data as possible. Let's go with conversion path.